File: admin/src/lib/g-launcher/font-catalogue.ts
/**
 * The families the font pickers offer, and the two the APK already carries.
 *
 * ─── GENERATED. DO NOT HAND-EDIT. ───────────────────────────────────────────
 *
 * Written by `apps/g_launcher/tools/gen_font_catalogue.py`, which emits this
 * file and `apps/g_launcher/assets/fonts/catalogue.json` in the same run.
 *
 * ONE SOURCE ON PURPOSE. Two hand-kept lists drift, and the drift is silent in
 * the worst direction: the panel offers a family, the pack publishes, and the
 * device resolves a name nothing registered and paints Roboto. Nothing errors
 * anywhere, and you find out from a screenshot.
 */

export type FontCategory = 'sans-serif' | 'serif' | 'display' | 'monospace';

export interface FontEntry {
  family: string;
  category: FontCategory;
}

/**
 * Declared in `pubspec.yaml`, so they resolve on a cold boot with no network.
 *
 * ─── `UbuntuMono` IS NOT `Ubuntu Mono` ──────────────────────────────────────
 *
 * The first is this bundled family. The second is the Google Fonts name for a
 * different copy of the same typeface, which the device fetches at runtime.
 * They are different strings, they take different code paths, and swapping one
 * for the other would cost the trust-anchor distro its offline rendering for no
 * visible gain. The picker therefore lists bundled families separately and
 * never rewrites one spelling to the other.
 */
export const BUNDLED_FONTS: FontEntry[] = [
  { family: 'Ubuntu', category: 'sans-serif' },
  { family: 'UbuntuMono', category: 'monospace' },
];

/** Fetched on device through `google_fonts`. Sorted at generation time. */
export const GOOGLE_FONTS: FontEntry[] = [
  { family: "Anonymous Pro", category: "monospace" },
  { family: "Anton", category: "display" },
  { family: "Archivo", category: "sans-serif" },
  { family: "Assistant", category: "sans-serif" },
  { family: "Audiowide", category: "display" },
  { family: "Azeret Mono", category: "monospace" },
  { family: "Barlow", category: "sans-serif" },
  { family: "Bebas Neue", category: "display" },
  { family: "Bitter", category: "serif" },
  { family: "Cabin", category: "sans-serif" },
  { family: "Comfortaa", category: "display" },
  { family: "Cormorant Garamond", category: "serif" },
  { family: "Courier Prime", category: "monospace" },
  { family: "Cousine", category: "monospace" },
  { family: "Crimson Text", category: "serif" },
  { family: "DM Mono", category: "monospace" },
  { family: "DM Sans", category: "sans-serif" },
  { family: "Domine", category: "serif" },
  { family: "EB Garamond", category: "serif" },
  { family: "Exo 2", category: "sans-serif" },
  { family: "Figtree", category: "sans-serif" },
  { family: "Fira Code", category: "monospace" },
  { family: "Fira Mono", category: "monospace" },
  { family: "Fira Sans", category: "sans-serif" },
  { family: "Fredoka", category: "display" },
  { family: "Heebo", category: "sans-serif" },
  { family: "IBM Plex Mono", category: "monospace" },
  { family: "IBM Plex Sans", category: "sans-serif" },
  { family: "IBM Plex Serif", category: "serif" },
  { family: "Inconsolata", category: "monospace" },
  { family: "Inter", category: "sans-serif" },
  { family: "JetBrains Mono", category: "monospace" },
  { family: "Josefin Sans", category: "sans-serif" },
  { family: "Karla", category: "sans-serif" },
  { family: "Lato", category: "sans-serif" },
  { family: "Lexend", category: "sans-serif" },
  { family: "Libre Baskerville", category: "serif" },
  { family: "Lora", category: "serif" },
  { family: "Manrope", category: "sans-serif" },
  { family: "Martian Mono", category: "monospace" },
  { family: "Merriweather", category: "serif" },
  { family: "Montserrat", category: "sans-serif" },
  { family: "Mulish", category: "sans-serif" },
  { family: "Nanum Gothic Coding", category: "monospace" },
  { family: "Newsreader", category: "serif" },
  { family: "Noto Sans", category: "sans-serif" },
  { family: "Noto Sans Mono", category: "monospace" },
  { family: "Noto Serif", category: "serif" },
  { family: "Nunito", category: "sans-serif" },
  { family: "Nunito Sans", category: "sans-serif" },
  { family: "Open Sans", category: "sans-serif" },
  { family: "Orbitron", category: "display" },
  { family: "Oswald", category: "display" },
  { family: "Outfit", category: "sans-serif" },
  { family: "Overpass Mono", category: "monospace" },
  { family: "Oxygen", category: "sans-serif" },
  { family: "PT Sans", category: "sans-serif" },
  { family: "PT Serif", category: "serif" },
  { family: "Playfair Display", category: "serif" },
  { family: "Plus Jakarta Sans", category: "sans-serif" },
  { family: "Poppins", category: "sans-serif" },
  { family: "Public Sans", category: "sans-serif" },
  { family: "Quicksand", category: "sans-serif" },
  { family: "Raleway", category: "sans-serif" },
  { family: "Red Hat Display", category: "sans-serif" },
  { family: "Red Hat Mono", category: "monospace" },
  { family: "Righteous", category: "display" },
  { family: "Roboto", category: "sans-serif" },
  { family: "Roboto Mono", category: "monospace" },
  { family: "Roboto Slab", category: "serif" },
  { family: "Rubik", category: "sans-serif" },
  { family: "Share Tech Mono", category: "monospace" },
  { family: "Sora", category: "sans-serif" },
  { family: "Source Code Pro", category: "monospace" },
  { family: "Source Sans 3", category: "sans-serif" },
  { family: "Source Serif 4", category: "serif" },
  { family: "Space Grotesk", category: "sans-serif" },
  { family: "Space Mono", category: "monospace" },
  { family: "Spectral", category: "serif" },
  { family: "Titillium Web", category: "sans-serif" },
  { family: "Ubuntu", category: "sans-serif" },
  { family: "Ubuntu Mono", category: "monospace" },
  { family: "Urbanist", category: "sans-serif" },
  { family: "Work Sans", category: "sans-serif" },
  { family: "Zilla Slab", category: "serif" },
];

/** True for a family the device can resolve without the pack shipping files. */
export function isKnownFamily(family: string): boolean {
  const f = family.trim();
  if (!f) return false;
  return (
    BUNDLED_FONTS.some((e) => e.family === f) ||
    GOOGLE_FONTS.some((e) => e.family === f)
  );
}

/**
 * Fixed-advance, as far as the catalogue knows.
 *
 * A bundled family is answered from `BUNDLED_FONTS`, a catalogue family from its
 * category, and anything else returns TRUE rather than false. That default is
 * deliberate: an unknown family is one the pack ships itself, and the panel has
 * no way to inspect a TTF's advance widths. Guessing "not monospaced" there
 * would block a legitimate publish over something it cannot actually check.
 */
export function isMonospaceFamily(family: string): boolean {
  const f = family.trim();
  const bundled = BUNDLED_FONTS.find((e) => e.family === f);
  if (bundled) return bundled.category === 'monospace';
  const google = GOOGLE_FONTS.find((e) => e.family === f);
  if (google) return google.category === 'monospace';
  return true;
}

export function isBundledFamily(family: string): boolean {
  return BUNDLED_FONTS.some((e) => e.family === family.trim());
}

/**
 * What a picker shows.
 *
 * The mono list is filtered to fixed-advance families, and that gate is not
 * cosmetic: `terminal_screen.dart` derives the PTY column count by measuring a
 * run of glyphs in this family. A proportional face makes the count too
 * generous, the remote host formats for a width the screen does not have, and
 * its output wraps mid-field.
 *
 * `Ubuntu` is therefore absent from the mono list even though it is bundled.
 */
export function fontOptions(mono: boolean): {
  bundled: FontEntry[];
  google: FontEntry[];
} {
  const keep = (e: FontEntry) => (mono ? e.category === 'monospace' : true);
  return {
    bundled: BUNDLED_FONTS.filter(keep),
    google: GOOGLE_FONTS.filter(keep),
  };
}

/**
 * The sample drawn in each row.
 *
 * A display name answers its own question: seeing "Playfair Display" set in
 * Playfair Display is the whole decision. A monospace name answers almost
 * nothing, because what is being chosen there is whether zero is
 * distinguishable from capital O, and whether the family ligates. Fira Code and
 * Fira Mono are the same typeface except for the ligatures.
 */
export function fontSample(family: string, mono: boolean): string {
  return mono ? `${family}  0O1lI  => != ===` : family;
}

/**
 * The stylesheet that makes the previews real, built once for the whole list.
 *
 * Weights 400 and 700 only, matching what the device fetches, so a preview
 * cannot show a weight the phone will not have. Bundled families are excluded:
 * asking Google Fonts for `UbuntuMono` returns nothing, and the row falls back
 * to the panel's own face, which is the honest result for a family that lives
 * in the APK rather than on a CDN.
 */
export function googleFontsHref(): string {
  const families = GOOGLE_FONTS.map(
    (e) => `family=${encodeURIComponent(e.family).replace(/%20/g, '+')}:wght@400;700`,
  ).join('&');
  return `https://fonts.googleapis.com/css2?${families}&display=swap`;
}

File: apps/g_launcher/lib/engine/font_catalogue.dart
/// The families a user can choose from, and what the two special choices mean.
///
/// ─── THREE STATES, AND ONLY ONE OF THEM IS A FAMILY NAME ────────────────────
///
/// A stored override is a `String?` on [LauncherPrefs], and it carries three
/// meanings:
///
///   null                the distro's own font. Kali stays in Kali's face.
///   [systemChoice]      the platform typeface, whatever the phone is set to.
///   anything else       that family, fetched from the Play Services provider.
///
/// [systemChoice] is a real stored value rather than a second null, because
/// "follow the phone" and "follow the distro" are different answers and a single
/// null cannot hold both. It resolves to a null `fontFamily`, which is how
/// Flutter spells "use the platform default".
///
/// [systemMonoChoice] is NOT the same thing and the difference is load-bearing.
/// The platform default is proportional, and `terminal_screen.dart` derives the
/// PTY column count by measuring a run of glyphs in the mono family. A
/// proportional face there does not just look wrong: the column count goes out
/// too generous, the remote host formats for a width the screen does not have,
/// and its own output wraps mid-field. So the mono picker offers Android's
/// `monospace` alias instead, which the platform resolves to the device's own
/// fixed-advance face.
///
/// ─── VENDORED, NOT FETCHED ──────────────────────────────────────────────────
///
/// The list ships in the APK. It could have come down the CDN beside the pack
/// index, signed the same way, and that was considered and dropped: adding a
/// family is not urgent enough to be worth a second signed channel and a native
/// fetch path, and a curated eighty-five scrolls better than the full library.
/// Move it to the CDN the first time a family cannot wait for a release.
library;

import 'dart:convert';

import 'package:flutter/services.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// Families declared in `pubspec.yaml` and therefore present in the APK.
///
/// ─── WHY THIS LIST HAS TO EXIST ─────────────────────────────────────────────
///
/// Everything else a theme can name is fetched at runtime. These two are not,
/// and asking `google_fonts` for them is worse than pointless: `UbuntuMono` is
/// not a Google Fonts family at all (that one is spelled `Ubuntu Mono`), so the
/// request fails, and `Ubuntu` would fetch a second copy of a typeface already
/// sitting in the APK.
///
/// The launcher's default distro must also paint correctly on a first cold boot
/// with no network, which is exactly the case a fetch cannot serve.
const Set<String> bundledFamilies = <String>{'Ubuntu', 'UbuntuMono'};

/// Follow the phone's own typeface. See the note above on why this is not null.
const String systemChoice = 'system';

/// Follow the phone's own MONOSPACE typeface. `monospace` is an Android font
/// alias, not a family this app ships, and the platform resolves it.
const String systemMonoChoice = 'monospace';

/// True when [choice] names something the platform resolves rather than
/// something that has to be fetched and registered.
bool isPlatformChoice(String? choice) =>
    choice == systemChoice || choice == systemMonoChoice;

/// One family the picker can offer.
class FontEntry {
  const FontEntry({
    required this.family,
    required this.category,
    required this.licence,
  });

  final String family;

  /// `sans-serif`, `serif`, `display` or `monospace`. The mono picker filters on
  /// this, which is the only thing standing between a proportional face and the
  /// column-count bug described above.
  final String category;

  /// `ofl`, `apache` or `ufl`. Selects which text is registered with
  /// [LicenseRegistry] when this family is actually loaded; see `main.dart`.
  final String licence;

  bool get isMono => category == 'monospace';

  /// What the picker draws in this family.
  ///
  /// ─── THE SAMPLE BELONGS TO THE PICKER, NOT TO THE FAMILY ────────────────
  ///
  /// This took [mono] as a parameter after a first cut keyed it off the
  /// family's own category, which put the `0O1lI => != ===` strip on Anonymous
  /// Pro and Azeret Mono in the DISPLAY list, where it means nothing and is
  /// simply noise in a column of names.
  ///
  /// A monospace family appearing in the display picker is a perfectly ordinary
  /// choice: someone can want their app labels in a typewriter face. What they
  /// are deciding there is what the letters look like, and the name answers it.
  ///
  /// ─── WHY THE MONO SAMPLE IS LONGER ──────────────────────────────────────
  ///
  /// In the mono picker the name answers almost nothing. Someone choosing a
  /// terminal face is deciding two things that are invisible in the word
  /// itself: whether zero is distinguishable from capital O and one from lower
  /// case l, and whether the family ligates `=>` and `!=` into arrows. Fira Code
  /// and Fira Mono are the same typeface except for the ligatures, so without
  /// that strip the picker cannot tell them apart at all.
  String sampleFor({required bool mono}) =>
      mono ? '$family  0O1lI  => != ===' : family;

  static FontEntry fromJson(Map<String, dynamic> j) => FontEntry(
        family: (j['family'] as String?) ?? '',
        category: (j['category'] as String?) ?? 'sans-serif',
        licence: (j['licence'] as String?) ?? 'ofl',
      );
}

/// Loaded once and held, because the picker can be opened repeatedly and the
/// file does not change between openings.
class FontCatalogue {
  const FontCatalogue._(this.families);

  final List<FontEntry> families;

  static FontCatalogue? _cached;

  /// Every family, sorted as authored (alphabetical, done at generation time).
  static Future<FontCatalogue> load() async {
    final cached = _cached;
    if (cached != null) return cached;

    try {
      final raw = await rootBundle.loadString('assets/fonts/catalogue.json');
      final doc = jsonDecode(raw) as Map<String, dynamic>;
      final list = ((doc['families'] as List?) ?? const [])
          .whereType<Map<String, dynamic>>()
          .map(FontEntry.fromJson)
          .where((e) => e.family.isNotEmpty)
          .toList(growable: false);

      return _cached = FontCatalogue._(list);
    } catch (e) {
      // A missing or malformed catalogue means the picker offers the two
      // platform choices and nothing else, which is a thin picker rather than a
      // broken screen. NOT cached, so a transient read failure can recover.
      return const FontCatalogue._(<FontEntry>[]);
    }
  }

  /// Only the fixed-advance families. What the mono picker shows.
  List<FontEntry> get monospace =>
      families.where((e) => e.isMono).toList(growable: false);

  /// The licence id for [family], or null when it is not one of ours: a pack's
  /// own font, a platform choice, or a stale stored value from a catalogue that
  /// has since dropped it.
  String? licenceFor(String family) {
    for (final e in families) {
      if (e.family == family) return e.licence;
    }
    return null;
  }
}

/// What the settings rows watch.
///
/// A [FutureProvider] because a settings section builder returns widgets
/// synchronously and cannot await an asset read. It renders with an empty list
/// for the frame it takes to parse and rebuilds with the families, which nobody
/// sees and which is far simpler than making the whole section async.
final fontCatalogueProvider =
    FutureProvider<FontCatalogue>((ref) => FontCatalogue.load());
